### DISCUSSION TITLE
Does anyone have a map of the code, or a guide to how to follow it?

### DISCUSSION THREAD

**Discussion by Zitchas** [endless-sky/endless-sky#7981] [Balance, Mechanics, and Interactions]
I'm trying to learn more of the source code side of the game, and it is pretty hard to follow. I'm not sure even where to begin. 

Does anyone have any maps, flowcharts, or guides as to how things connect together?

**Comment by RisingLeaf** [comment]
I asked the same question when I started coding for this game, at this time the answer was no.

Something I proposed in this PR is creating parseable comments in all headers to be able to auto generate a documentation. But that is far from being finished.

What I suggest you is trying to do some simple bug fixes to get more familiar with the code and then slowly get to more complex problems.

**Comment by vitalchip** [comment]
Some of it is straightforward and some a bit more convoluted. All I can really offer is to say that I'd never seen C++ until I started poking around this game, other than some basic on a Ti99 back in 83 and about 20 lines of 6502 assembler in 87 I've never 'coded'.

I read it, changed some stuff, copied one bit to another and while I still can't 'code' I can get stuff done, so have faith, come back to a puzzle the next day and you will be surprised how much you can do.

**Reply by Zitchas** [reply]
I've started a branch where I'm poking at some stuff to do with radar colors. Would you be interested in taking a look and seeing if I'm on the right track please? I made a PR for ease of seeing the changes and discussion.

https://github.com/Zitchas/endless-sky/pull/130

Edit, it was number 129, but I had to redo the PR to get rid of some unwanted stuff, so now it's 130.

**Reply by Amazinite** [reply]
Basically how I started. There are sections of the codebase that by themselves aren't too difficult to understand. From there it's just a matter of recognizing similar patterns of how the game is coded in different files and slowly expanding your understanding of things.

**Comment by Amazinite** [comment]
I have on my todo list a task to create a short video series explaining in high level some of the more important parts of the source code as they pertain to content creation. No idea when I'll be able to get to it, but it'd hopefully help with questions like this. I'd say the easiest entry point though is to take a look at UniverseObject::LoadFile. This is where each file from the data folder is loaded into the game and parsed into usable information. It turns the file path into a DataFile, which turns a file into a tree of DataNodes. Each root node (a line with no indentation before it, e.g. "mission," "ship," "outfit") is then parsed by its own class (or if it's simple enough, UniverseObject does the parsing itself, like for Color or Star). Looking at the big if/else tree in LoadFile, you can see each root node that the game recognizes. You can then go to the class for each root node to see how it parses its own information. Some root nodes are pretty simple, like Hazard, while others are pretty complex, like Ship or Mission.

**Reply by Zitchas** [reply]
Thanks for the starting point! 
I've been doing the "Find All" in the source folder for keywords to try to look for relevant stuff method so far.

**Reply by warp-core** [reply]
> I've been doing the "Find All" in the source folder for keywords to try to look for relevant stuff method so far.

Don't expect that to stop being useful; I do it all the time.

**Reply by Amazinite** [reply]
I will add that UniverseObject and the stuff it loads isn't everything. PlayerInfo is another important class, which is the actual player save that you're currently running. Engine and AI are another two big ones, where Engine is what actually does most of the running of things frame by frame, and AI being what controls ships. But those are what I'd consider the top four classes to take a look at, in increasing order of difficulty/complexity: UniverseObject (and by extension GameData, as they're closely linked classes right now), PlayerInfo, Engine, and AI. If you can wrap your head around each of those, you're probably on the right track, as almost everything else cascades down from those four.

**Reply by Zitchas** [reply]
Thanks!
